### PR TITLE
Parser: calculating interface hash only for primary files.

### PR DESCRIPTION
**Pull request by nkcsgexi**
Per discussion we had in https://github.com/apple/swift/pull/19192, we only need to collect interface hashes for primary files. This change allows parser to consume tokens out of order for non-primary files.

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: include/swift/AST/Module.h [hunk 1]
```diff
@@ -1117,20 +1118,23 @@ class SourceFile final : public FileUnit {
   /// Set the root refinement context for the file.
   void setTypeRefinementContext(TypeRefinementContext *TRC);
 
+  void initInterfaceHashIfEmpty() {
```
Comment (jrose-apple) [inline include/swift/AST/Module.h]: This doesn't seem right to me. I'd expect something like `enableInterfaceHash()`, which would assert if it were already enabled.

File: include/swift/Subsystems.h [hunk 2]
```diff
@@ -117,6 +117,7 @@ namespace swift {
   ///
   /// \return true if the parser found code with side effects.
   bool parseIntoSourceFile(SourceFile &SF, unsigned BufferID, bool *Done,
+                           bool CollectInterfaceHash,
```
Comment (jrose-apple) [inline include/swift/Subsystems.h]: …and then you wouldn't need this.

**Review by jrose-apple — COMMENTED** [review]
Glad it's looking easy! I think I'd prefer a different implementation strategy though…it feels like you mixed two of my ideas.

File: include/swift/AST/Module.h (hunk 1 above)
Comment (nkcsgexi) [reply include/swift/AST/Module.h]: That makes sense.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Frontend/Frontend.cpp [hunk 3]
```diff
@@ -745,7 +745,8 @@ void CompilerInstance::parseLibraryFile(
   auto &Diags = NextInput->getASTContext().Diags;
   auto DidSuppressWarnings = Diags.getSuppressWarnings();
   Diags.setSuppressWarnings(DidSuppressWarnings || !IsPrimary);
-
+  if (IsPrimary)
+    NextInput->enableInterfaceHash();
```
Comment (jrose-apple) [inline lib/Frontend/Frontend.cpp]: I think you can simplify this by doing it even earlier, in `CompilerInstance::createSourceFileForMainModule`.

Comment (nkcsgexi) [reply lib/Frontend/Frontend.cpp]: I've put it in `recordPrimarySourceFile`.

**Review by jrose-apple — APPROVED** [review]
Now that's a nice, small diff! Thanks for working on this, Xi. Hopefully it still gives you the speedups you need. (And it'll be something you can refine later, I guess.)

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test OS X Platform

**Review by rintaro — APPROVED** [review]
Thank you Xi! This patch itself may speedups non-primary files parsing.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Pull request merged**